### PR TITLE
Block syntax for associations.

### PR DESCRIPTION
**Pull request by beauby**
This PR is half of #1262, introducing the block syntax for overriding associations (and removing the legacy method defining syntax).

Currently 2 tests fail, I believe those tests are ill (they seem to expect an association to be treated as an attribute).

### CODE REVIEW

File: lib/active_model/serializer/associations.rb [hunk 1]
```diff
@@ -29,7 +28,8 @@ class << base
 
       module ClassMethods
         def inherited(base)
-          base._reflections = self._reflections.try(:dup) || []
```
Comment (NullVoxPopuli) [inline lib/active_model/serializer/associations.rb]: why was this a `try` before?

File: lib/active_model/serializer/reflection.rb [hunk 2]
```diff
@@ -54,7 +61,7 @@ def build_association(subject, parent_serializer_options)
             reflection_options[:virtual_value] = association_value.try(:as_json) || association_value
           end
         elsif !association_value.nil? && !association_value.instance_of?(Object)
-          reflection_options[:virtual_value] = association_value
+          reflection_options[:virtual_value] = association_value.try(:as_json) || association_value
```
Comment (NullVoxPopuli) [inline lib/active_model/serializer/reflection.rb]: why is there a try here?

File: test/fixtures/poro.rb [hunk 3]
```diff
@@ -212,13 +208,12 @@ def self.root_name
 VirtualValueSerializer = Class.new(ActiveModel::Serializer) do
   attributes :id
 
-  has_many :reviews, virtual_value: [{ id: 1 }, { id: 2 }]
-  has_one :maker, virtual_value: { id: 1 }
-
-  def reviews
+  has_many :reviews do
```
Comment (NullVoxPopuli) [inline test/fixtures/poro.rb]: much cleaner than `virtual_value` :-)

File: lib/active_model/serializer/associations.rb (hunk 1 above)
Comment (beauby) [inline lib/active_model/serializer/associations.rb]: Because the `self._reflections` was not initialized in the constructor.

File: lib/active_model/serializer/reflection.rb (hunk 2 above)
Comment (beauby) [inline lib/active_model/serializer/reflection.rb]: For consistency with the way we handle collections for which we can't find serializers.

File: test/fixtures/poro.rb (hunk 3 above)
Comment (beauby) [inline test/fixtures/poro.rb]: Yeah, plus I'm not sure `virtual_value` was officially part of the public API.

File: lib/active_model/serializer/reflection.rb (hunk 2 above)
Comment (NullVoxPopuli) [inline lib/active_model/serializer/reflection.rb]: how do you mean? what scenarios are those?

Comment (beauby) [inline lib/active_model/serializer/reflection.rb]: When doing 
```ruby
has_many :blah do 
  [ { id: 3, type: 'blah' }, {...} ]
end
```
or
```ruby
belongs_to :blah do
  { id: blah, type: 'blah' }
end
```

File: lib/active_model/serializer/associations.rb (hunk 1 above)
Comment (NullVoxPopuli) [inline lib/active_model/serializer/associations.rb]: so any class that inherits `Associations` must have a `_reflections` accessor? 
can you document that, please? 
I know it's outside the scope of a refactor, but since we have next to none of this type of documentation, I think it will be good to do it as we go.

File: lib/active_model/serializer/reflection.rb (hunk 2 above)
Comment (NullVoxPopuli) [inline lib/active_model/serializer/reflection.rb]: can you write some comments about why try is being used and a quick blurb about when association_value might not have `as_json`

thanks ;-)

File: test/fixtures/poro.rb (hunk 3 above)
Comment (NullVoxPopuli) [inline test/fixtures/poro.rb]: even better!

File: lib/active_model/serializer/associations.rb (hunk 1 above)
Comment (beauby) [inline lib/active_model/serializer/associations.rb]: No, including `Associations` *does* declare a `_reflections` accessor.

Comment (beauby) [inline lib/active_model/serializer/associations.rb]: c.f. https://github.com/rails-api/active_model_serializers/pull/1311/files#diff-b2b037951247cfa436c1f2201f5c12dfL16

Comment (NullVoxPopuli) [inline lib/active_model/serializer/associations.rb]: I see below on line 32, that `base._reflections` is set to a dup of `_reflections` there does the duped `_reflections` come from?

Comment (beauby) [inline lib/active_model/serializer/associations.rb]: That's upon inheritance, so it is duped from the parent class's

Comment (NullVoxPopuli) [inline lib/active_model/serializer/associations.rb]: oh ok, so you're setting `base._reflections` just to an empty array?

Comment (beauby) [inline lib/active_model/serializer/associations.rb]: depending if the parent class has some reflections already or not

Comment (NullVoxPopuli) [inline lib/active_model/serializer/associations.rb]: I'm confused. Sorry. Can you summarize?

Comment (beauby) [inline lib/active_model/serializer/associations.rb]: Module `Associations` defines (on the class that includes it) a class attribute `_reflections` that is initialized to `[]` and is inherited.

File: lib/active_model/serializer/reflection.rb (hunk 2 above)
Comment (beauby) [inline lib/active_model/serializer/reflection.rb]: `association_value` might be user-supplied and it might be anything

File: lib/active_model/serializer/associations.rb (hunk 1 above)
Comment (beauby) [inline lib/active_model/serializer/associations.rb]: To be clear, those changes are mainly (if not only) cosmetic, in order to be consistent with how we deal with inheritance of class attributes all around, they're not necessary for this PR.

File: lib/active_model/serializer/reflection.rb [hunk 4]
```diff
@@ -17,7 +17,21 @@ class Serializer
     #
     # So you can inspect reflections in your Adapters.
```
Comment (NullVoxPopuli) [inline lib/active_model/serializer/reflection.rb]: what is a reflection?

Comment (beauby) [inline lib/active_model/serializer/reflection.rb]: https://github.com/rails-api/active_model_serializers/blob/master/lib/active_model/serializer/reflection.rb

Comment (NullVoxPopuli) [inline lib/active_model/serializer/reflection.rb]: oh. ha. didn't see there was more. thanks

**Comment by beauby** [comment]
Should be green.

File: lib/active_model/serializer/associations.rb [hunk 5]
```diff
@@ -13,9 +13,8 @@ module Associations
       DEFAULT_INCLUDE_TREE = ActiveModel::Serializer::IncludeTree.from_string('*')
 
       included do |base|
-        class << base
-          attr_accessor :_reflections
-        end
+        base.class_attribute :_reflections
```
Comment (bf4) [inline lib/active_model/serializer/associations.rb]: I don't think we want this to be instance readable/writable? Do we want these to be inheritable?

Comment (bf4) [inline lib/active_model/serializer/associations.rb]: re: inhertiable, yes, oops 

```diff
-          base._reflections = self._reflections.try(:dup) || []
+          super
+          base._reflections = _reflections.dup
```

File: lib/active_model/serializer/associations.rb [hunk 6]
```diff
@@ -29,58 +28,44 @@ class << base
...
         # @param [Symbol] name of the association
         # @param [Hash<Symbol => any>] options for the reflection
+        # @param [Block] optional block to customize value of the reflection
         # @return [void]
         #
         # @example
         #  has_many :comments, serializer: CommentSummarySerializer
         #
-        def has_many(name, options = {})
-          associate HasManyReflection.new(name, options)
```
Comment (bf4) [inline lib/active_model/serializer/associations.rb]: can we keep the `associate` method?  I don't feel strongly about it, but I like it...

File: lib/active_model/serializer/reflection.rb [hunk 7]
```diff
@@ -17,7 +17,7 @@ class Serializer
...
     #
-    Reflection = Struct.new(:name, :options) do
+    Reflection = Struct.new(:name, :options, :proc) do
```
Comment (bf4) [inline lib/active_model/serializer/reflection.rb]: I prefer `callable` to refer to a `callable` when, like here, it can be a block, proc, or lambda

File: lib/active_model/serializer/reflection.rb [hunk 8]
```diff
@@ -40,7 +40,11 @@ class Serializer
       # @api private
       #
       def build_association(subject, parent_serializer_options)
-        association_value = subject.send(name)
+        association_value = if proc.nil?
+                              subject.object.send(name)
+                            else
+                              subject.instance_eval(&proc)
```
Comment (bf4) [inline lib/active_model/serializer/reflection.rb]: Do we need a design that requires us to instance_eval?

File: lib/active_model/serializer/reflection.rb (hunk 2 above)
Comment (bf4) [inline lib/active_model/serializer/reflection.rb]: I don't think we need the `try(:as_json)` change for this pr, or in general

File: test/fixtures/poro.rb [hunk 9]
```diff
@@ -62,12 +62,10 @@ module Spam; end
   attributes :id, :title, :body
 
   has_many :comments
-  belongs_to :blog
-  belongs_to :author
-
-  def blog
+  belongs_to :blog do
     Blog.new(id: 999, name: 'Custom blog')
```
Comment (bf4) [inline test/fixtures/poro.rb]: What does this mean?

Comment (bf4) [inline test/fixtures/poro.rb]: n/m it just looks weird to create an artificial association, but that's pre-existing

**Comment by bf4** [comment]
Closed by https://github.com/rails-api/active_model_serializers/issues/1356
